merkle/src/hash.rs

```diff
@@ -0,0 +1,78 @@
+use ark_ff::{BigInteger, PrimeField};
+use blake2::{Blake2s256, Digest};
+use felt::Felt252;
+use poseidon::{FieldHasher, Poseidon3};
+use sha3::Keccak256;
+use std::{fmt::Debug, marker::PhantomData};
+
+pub trait Hasher<F: PrimeField> {
+    type Output: Clone + Eq + Default + Debug;
+
+    // compress a list of internal nodes into a single internal node
+    fn node(input: &[Self::Output]) -> Self::Output;
+
+    // compress a list of leaves into a single leaf
+    fn leaf(input: &[F]) -> Self::Output;
+}
+
+pub struct Blake2s256Hasher<F: PrimeField> {
+    _ph: PhantomData<F>,
+}
+
+impl<F: PrimeField> Hasher<F> for Blake2s256Hasher<F> {
```

Comment:
I wonder if he can have a generic implementation for every `Digest`.

Reply:
I think it could be possible for `Blake2s256Hasher` and `Keccak256Hasher`. But I am not sure if it can be done for `Poseidon3Hasher` because its implementation of the `Hasher` trait slightly differs from the other two.